### PR TITLE
feat: Implement logfmt (Heroku) formatted log output

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/516

# Rationale
The heroku logfmt logging format is commonly used (including The InfluxDB cloud service). The Rust standard log format is less commonly used (at the time of this writing). Thus, IOx should make it easy to hook up logs to existing systems.

I *really* want structured logs to appear in our internal InfluxData testing tools clusters as we roll IOx out. logfmt seems like the shortest path to get there.

## Why roll our own?
I could find nothing suitable: https://crates.io/search?q=logfmt

I did not go with the seemingly obvious choice, https://crates.io/crates/logfmt_logger (though thanks @mcountryman for creating that), because
1. Doesn't (obviously) work with `tracing_subscriber`
2. Has been downloaded 65 (not a typo) times total ever
3. Doesn't seem to support fields other than `target` and `message` - https://github.com/mcountryman/logfmt_logger
4. The examples seem to be different than https://brandur.org/logfmt#implementations shows (e.g. the use of "message" instead of "msg" as a field name, for example)

# Changes
1. A `tracing_subscriber` Layer that implements what I percieve as the appropriate `log_fmt` formatting rules
2. Full coverage unit and integration tests that demonstrate the behavior
3. A new command line `--log_format` and environment variable `INFLUXDB_IOX_LOG_FORMAT` that can be used to choose what format is used (defaults to the Rust style)

## Why its own crate?
1. It made the test/recompile cycle faster
2. It makes it easier to open source this / contribute it back to the tracing_subscriber crate in the future if we want

# Examples:

## `influxdb_iox` with logfmt formatted logs:

```
cd /Users/alamb/Software/influxdb_iox && INFLUXDB_IOX_LOGFORMAT="rust" cargo run -- server --log_format=logfmt
   Compiling influxdb_iox v0.1.0 (/Users/alamb/Software/influxdb_iox)
    Finished dev [unoptimized + debuginfo] target(s) in 18.16s
     Running `target/debug/influxdb_iox server --log_format=logfmt`
level=warn msg="NO PERSISTENCE: using memory for object storage" target="influxdb_iox::influxdb_ioxd" location="src/influxdb_ioxd.rs:108" time=1612181728438978000
level=warn msg="server ID not set. ID must be set via the INFLUXDB_IOX_ID config or API before writing or querying data." target="influxdb_iox::influxdb_ioxd" location="src/influxdb_ioxd.rs:127" time=1612181728439095000
level=info msg="gRPC server listening" bind_address=127.0.0.1:8082 target="influxdb_iox::influxdb_ioxd" location="src/influxdb_ioxd.rs:139" time=1612181728439293000
level=info msg="HTTP server listening" bind_address=127.0.0.1:8080 target="influxdb_iox::influxdb_ioxd" location="src/influxdb_ioxd.rs:149" time=1612181728447419000
InfluxDB IOx server ready
```

## `influxdb_iox` default behavior / Rust formatted logs (unchanged)

```
cd /Users/alamb/Software/influxdb_iox && cargo run -- server
    Blocking waiting for file lock on build directory
   Compiling influxdb_iox v0.1.0 (/Users/alamb/Software/influxdb_iox)
    Finished dev [unoptimized + debuginfo] target(s) in 21.20s
     Running `target/debug/influxdb_iox server`
Jan 31 13:19:39.059  WARN influxdb_iox::influxdb_ioxd: NO PERSISTENCE: using memory for object storage
Jan 31 13:19:39.060  WARN influxdb_iox::influxdb_ioxd: server ID not set. ID must be set via the INFLUXDB_IOX_ID config or API before writing or querying data.
Jan 31 13:19:39.060  INFO influxdb_iox::influxdb_ioxd: gRPC server listening bind_address=127.0.0.1:8082
Jan 31 13:19:39.068  INFO influxdb_iox::influxdb_ioxd: HTTP server listening bind_address=127.0.0.1:8080
InfluxDB IOx server ready
```

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: logfmt/tests/logging.rs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+            normalized_expected, normalized_actual,
+            "\n\nexpected:\n\n{:#?}\nactual:\n\n{:#?}\n\nnormalized_expected:\n\n{:#?}\nnormalized_actual:\n\n{:#?}\n\n",
+            expected_lines, actual_lines,
+            normalized_expected, normalized_actual
+        )
+    };
+}
+
+#[test]
+fn level() {
```
Comment (alamb) [inline logfmt/tests/logging.rs:38]: This is probably the best place to start to see the calls to `info!`, `error!` etc and the corresponding output

**Review by alamb — COMMENTED** [review]
The vast majority of this PR is tests. The actual code is just hooking up the `tracing_subscriber` plumbing and escaping things

image

**Comment by alamb** [comment]
Some notes:

making useful logs is a function of:
1. The actual formatting (dare I say serialization) of name/value pairs into a string
2. The conventions on the names of the fields and their meanings. 

This PR  implements the actual format .  @e-dard  notes (and I agree) that the most useful thing is to associate a single `trace_id` throughout all operations in a request. This PR does not do anything with trace_ids

**Comment by e-dard** [comment]
The most important thing I want to get out of this PR is the name of that computer game. Right on the tip of my tongue!

File: logfmt/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+        };
+
+        write!(writer, r#"level={}"#, level_str).ok();
+
+        Self { writer }
+    }
+
+    fn write_source_info(&mut self, event: &tracing::Event<'_>) {
+        let metadata = event.metadata();
+        write!(self.writer, " target=\"{}\"", metadata.target()).ok();
```
Comment (e-dard) [inline logfmt/src/lib.rs:102]: I was surprised to find that Rust doesn't have a format parameter for quoting the string (Go has `%q` for this). TIL.

File: logfmt/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    let c1 = value.chars().skip(1);
+    if c0.zip(c1).any(|(c0, c1)| c0 != '\\' && c1 == '"') {
+        return true;
+    }
+
+    if value.contains(' ') && !pre_quoted {
+        return true;
+    }
+
+    false
```
Comment (e-dard) [inline logfmt/src/lib.rs:239]: Alternatively:

```suggestion
    value.contains(' ') && !pre_quoted
```

File: logfmt/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+}
+
+/// escape any characters in name as needed, otherwise return string as is
+fn quote_and_escape(value: &'_ str) -> Cow<'_, str> {
```
Comment (e-dard) [inline logfmt/src/lib.rs:243]: 🐮

**Review by e-dard — APPROVED** [review]
This LGTM 👍  Thank you for putting the miles in to get this working 💯

File: logfmt/src/lib.rs (hunk 2 above)
Comment (domodwyer) [inline logfmt/src/lib.rs:102]: I think these values are user-controlled (but defaulted when using the event macros) so may need escaping?

**Review by domodwyer — COMMENTED** [review]
Looks cool but couple of questions!

I think we should publish this as a crate? I'm sure other people will want to use it!

**Comment by alamb** [comment]
Filed https://github.com/influxdata/influxdb_iox/issues/717 to track adding trace_id

File: logfmt/src/lib.rs (hunk 2 above)
Comment (alamb) [reply logfmt/src/lib.rs:102]: Sounds good -- will escape if needed

File: logfmt/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+                write!(self.writer, " module_path=\"{}\"", module_path).ok();
+            }
+        }
+        if let (Some(file), Some(line)) = (metadata.file(), metadata.line()) {
+            write!(self.writer, " location=\"{}:{}\"", file, line).ok();
+        }
+    }
+
+    fn write_span_id(&mut self, id: Id) {
+        write!(self.writer, " span={}", id.into_u64()).ok();
```
[the remark this answers is not in the document]
Comment (alamb) [reply logfmt/src/lib.rs:115]: Interestingly (and confusingly), spans don't actually seem to be propagated by the tracing_subscriber (there are events when spans get entered / exited but the layer itself I think is supposed to track them itself). I'll add a test case showing how this (doesn't work) in the end to end test

File: logfmt/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    #[test]
+    fn quote_and_escape_space_prequoted() {
+        assert_eq!(quote_and_escape("\"foo bar\""), "\"foo bar\"");
+    }
+
+    #[test]
+    fn quote_and_escape_space_prequoted_but_not_escaped() {
+        assert_eq!(quote_and_escape("\"foo \"bar\""), "\"\\\"foo \\\"bar\\\"\"");
```
[the remark this answers is not in the document]
Comment (alamb) [reply logfmt/src/lib.rs:307]: yeah, maybe -- the upside of this was that you get the same output you can copy/paste from the failure output when the tests run

File: logfmt/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+        if let (Some(file), Some(line)) = (metadata.file(), metadata.line()) {
+            write!(self.writer, " location=\"{}:{}\"", file, line).ok();
+        }
+    }
+
+    fn write_span_id(&mut self, id: Id) {
+        write!(self.writer, " span={}", id.into_u64()).ok();
+    }
+
+    fn write_timestamp(&mut self) {
```
[the remark this answers is not in the document]
Comment (alamb) [reply logfmt/src/lib.rs:118]: TLDR: not that I could find: https://docs.rs/tracing/0.1.22/tracing/struct.Event.html

However, I probably missed it

However I will say that I found the Traits and structs in `tracing` and `tracing_subscriber` non trivial to follow (partly due to the naming of things like a `Layer` struct that implements a `Layer` trait in a different module:  https://docs.rs/tracing-subscriber/0.2.15/tracing_subscriber/fmt/struct.Layer.html -- I was complaining about this to @e-dard and @pauldix ).

**Review by alamb — COMMENTED** [review]
Thanks @domodwyer  -- I will add some tests to demonstrate how span! works with this crate

**Comment by mcountryman** [comment]
> Looks cool but couple of questions!
> 
> I think we should publish this as a crate? I'm sure other people will want to use it!



> Looks cool but couple of questions!
> 
> I think we should publish this as a crate? I'm sure other people will want to use it!

I didn't get a chance to dig too far but, if this crate impls a [log](https://crates.io/crates/log) logger and you want to use the crate name [logfmt_logger](https://crates.io/crates/logfmt_logger), let me know.

File: logfmt/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (domodwyer) [reply logfmt/src/lib.rs:118]: Yeah I looked too - nothing! I guess from now on I'll view timestamps with a (slight) degree of skepticism!

File: logfmt/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (domodwyer) [reply logfmt/src/lib.rs:115]: Ah yes you're right - the span [Id docs](https://docs.rs/tracing/0.1.22/tracing/span/struct.Id.html) say:

>They are generated by Subscribers for each span as it is created

Which makes sense! I assumed it was a responsibility of the tracing framework, but it in retrospect that'd be a bit naff & restrictive 😆

[the remark this answers is not in the document]
Comment (alamb) [reply logfmt/src/lib.rs:115]: or "helpful" -- lol -- I'll clarify that this implementation doesn't (yet) do anything useful with span_ids (span fields in general actually)

**Comment by alamb** [comment]
Thanks @mcountryman  -- this crate implements a trait from the `tracing_subscriber` framework (which  implements its own logging subscriber, I think). I thing `logfmt_logger` without the `tracing_subscriber` bits will still be needed.

File: logfmt/tests/logging.rs [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+
+    assert_logs!(capture, expected);
+}
+
+#[test]
+fn event_multi_span() {
+    // Demonstrate the inclusion of span_id (as `span`)
+    let capture = CapturedWriter::new();
+
+    let span1 = span!(Level::INFO, "my_span", foo = "bar");
```
Comment (alamb) [inline logfmt/tests/logging.rs:168]: So I was messing around with spans based on this comment and I was able to get the span_id printed (what isn't propagated is the fields on the span definition) but I think that is ok. However, as @domodwyer  has pointed out these spans are likely not going to be all that helpful as they don't have the information needed to build a tree (yet)

**Comment by alamb** [comment]
I am happy to address any additional suggestions / changes in a follow up PR -- for now merging this one

**Pull request merged**